src/compiler/parsing/watf/string-literals.js

```diff
@@ -0,0 +1,82 @@
+// @flow
+
+// string literal characters cannot contain control codes
+const CONTROL_CODES = [
+  0, // null
+  7, // bell
+  8, // backspace
+  9, // horizontal
+  10, // line feed
+  11, // vertical tab
+  12, // form feed
+  13, // carriage return
+  26, // Control-Z
+  27, // escape
+  127 // delete
+];
+
+// escaped sequences can either be a two character hex value, or one of the
+// following single character codes
+function decodeControlCharacter(char: string) {
+  switch (char) {
+    case "t":
+      return 0x09;
+    case "n":
+      return 0x0a;
+    case "r":
+      return 0x0d;
+    case '"':
+      return 0x22;
+    case "′":
+      return 0x27;
+    case "\\":
+      return 0x5c;
+  }
+  return -1;
+}
+
+const ESCAPE_CHAR = 92; // backslash
+const QUOTE_CHAR = 34; // backslash
```

Comment:
Just FYI we use https://github.com/xtuc/charcodes/tree/master/packages/charcodes in Babel's parser. We can use it also here if it make it more clearer.